lectures/jax_intro.md

```diff
@@ -0,0 +1,600 @@
+
+
+### Installation
+
+JAX can be installed with or without GPU support by following [the install guide](https://github.com/google/jax).
+
+Note that JAX is pre-installed with GPU support on [Google Colab](https://colab.research.google.com/).
+
+If you do not have your own GPU, we recommend that you run this lecture on Colab.
+
++++
+
+## JAX as a NumPy Replacement
+
+
+One way to use JAX is as a plug-in NumPy replacement.  Let's look at the similarities and differences.
+
+### Similarities
+
+
+The following import is standard, replacing `import numpy as np`:
+
+```{code-cell} ipython3
+import jax
+import jax.numpy as jnp
+```
+
+Now we can use `jnp` in place of `np` for the usual array operations:
+
+```{code-cell} ipython3
+a = jnp.asarray((1.0, 3.2, -1.5))
+```
+
+```{code-cell} ipython3
+print(a)
+```
+
+```{code-cell} ipython3
+print(jnp.sum(a))
+```
+
+```{code-cell} ipython3
+print(jnp.mean(a))
+```
+
+```{code-cell} ipython3
+print(jnp.dot(a, a))
+```
+
+However, the array object `a` is not a NumPy array:
+
+```{code-cell} ipython3
+a
+```
+
+```{code-cell} ipython3
+type(a)
+```
+
+Even scalar-valued maps on arrays are of type `DeviceArray`:
+
+```{code-cell} ipython3
+jnp.sum(a)
+```
+
+The term `Device` refers to the hardware accelerator (GPU or TPU), although JAX falls back to the CPU if no accelerator is detected.
+
+(In the terminology of GPUs, the "host" is the machine that launches GPU operations, while the "device" is the GPU itself.)
+
+
+Operations on higher dimensional arrays is also similar to NumPy:
+
+```{code-cell} ipython3
+A = jnp.ones((2, 2))
+B = jnp.identity(2)
+A @ B
+```
+
+```{code-cell} ipython3
+from jax.numpy import linalg
+```
+
+```{code-cell} ipython3
+linalg.solve(B, A)
+```
+
+```{code-cell} ipython3
+linalg.eigh(B)  # Computes eigenvalues and eigenvectors
+```
+
+### Differences
+
+
+One difference between NumPy and JAX is that, when running on a GPU, JAX uses 32 bit floats by default.  
+
+This is standard for GPU computing and can lead to significant speed gains with small loss of precision.
+
+However, for some calculations precision matters.  In these cases 64 bit floats can be enforced via the command
+
+```{code-cell} ipython3
+jax.config.update("jax_enable_x64", True)
+```
+
+Let's check this works:
+
+```{code-cell} ipython3
+jnp.ones(3)
+```
+
+As a NumPy replacement, a more significant difference is that arrays are treated as **immutable**.  
+
+For example, with NumPy we can write 
+
+```{code-cell} ipython3
+import numpy as np
+a = np.linspace(0, 1, 3)
+a
+```
+
+and then mutate the data in memory:
+
+```{code-cell} ipython3
+a[0] = 1
+a
+```
+
+In JAX this fails:
+
+```{code-cell} ipython3
+a = jnp.linspace(0, 1, 3)
+a
+```
+
+```{code-cell} ipython3
+:tags: [raises-exception]
+a[0] = 1
+```
+
+In line with immutability, JAX does not support inplace operations:
+
+```{code-cell} ipython3
+a = np.array((2, 1))
+a.sort()
+a
+```
+
+```{code-cell} ipython3
+a = jnp.array((2, 1))
+a.sort()
+a
```

Comment:
```suggestion
a = jnp.array((2, 1))
a_new = a.sort()
a, a_new
```

This shows that `a_new` is a new sorted array but `a` still remains the same.

Reply:
Thanks @Smit-create ,  good idea.